Sources/TestingMacros/CMakeLists.txt

```diff
@@ -61,14 +48,53 @@ add_executable(TestingMacros
   TagMacro.swift
   TestDeclarationMacro.swift
   TestingMacrosMain.swift)
-set_target_properties(TestingMacros PROPERTIES
-  ENABLE_EXPORTS TRUE)
-target_compile_options(TestingMacros PRIVATE -parse-as-library)
-target_link_libraries(TestingMacros PRIVATE
-  SwiftSyntax::SwiftCompilerPlugin
+
+find_package(SwiftSyntax CONFIG GLOBAL)
+
+if(SwiftTesting_BuildMacrosAsExecutables)
+  # When building the macro plugin as an executable, clone and build
+  # swift-syntax.
+  include(FetchContent)
+  set(FETCHCONTENT_BASE_DIR ${CMAKE_BINARY_DIR}/_d)
+  # TODO: Update GIT_TAG to the 6.0 release tag once it is available.
+  FetchContent_Declare(SwiftSyntax
+    GIT_REPOSITORY https://github.com/swiftlang/swift-syntax
+    GIT_TAG 27b74edd5de625d0e399869a5af08f1501af8837)
+  FetchContent_MakeAvailable(SwiftSyntax)
+endif()
+
+# Include these modules _after_ swift-syntax is declared above, but _before_ the
+# macro plugin target is declared below, so that its settings are not applied to
+# the former but are applied to the latter.
+include(AvailabilityDefinitions)
+include(CompilerSettings)
+
+set(_TestingMacros_LinkedLibraries_PRIVATE
   SwiftSyntax::SwiftSyntax
   SwiftSyntax::SwiftSyntaxMacroExpansion
-  SwiftSyntax::SwiftSyntaxMacros
-  SwiftSyntax509
-  SwiftSyntax510
-  SwiftSyntax600)
+  SwiftSyntax::SwiftSyntaxMacros)
+
+if(SwiftTesting_BuildMacrosAsExecutables)
+  # When swift-syntax is built locally, the macro plugin must be built as an
+  # executable.
+  add_executable(TestingMacros ${_TestingMacros_SourceFiles})
+
+  set_target_properties(TestingMacros PROPERTIES
+    ENABLE_EXPORTS TRUE)
+
+  # Parse the module as a library, even though it's an executable, because it
+  # uses an `@main` type to define its entry point.
+  target_compile_options(TestingMacros PRIVATE -parse-as-library)
+
+  # Include the .swift file which contains its `@main` entry point type.
+  target_compile_definitions(TestingMacros PRIVATE SWT_NO_LIBRARY_MACRO_PLUGINS)
+
+  # Link the 'SwiftCompilerPlugin' target, but only when built as an executable.
+  list(APPEND _TestingMacros_LinkedLibraries_PRIVATE
+    SwiftSyntax::SwiftCompilerPlugin)
+else()
+  add_library(TestingMacros SHARED ${_TestingMacros_SourceFiles})
+endif()
```

Comment:
Sorry for the basic question, but can you elaborate on how/why it is problematic to store the file list in a variable? It would be pretty unfortunate if we repeated the list twice, and I worry it'll lead to us forgetting to keep them in sync and lead to maintenance issues. I'm trying to weigh that practical risk against the likelihood we'd actually encounter whatever problems this anti-pattern might have.

Reply:
The problem is that it encourages more imperative building of the code rather than declarative. The declarative model has been preferred since CMake 3.0 which is quite old at this point. In fact you see this problem already in the sense that you are writing imperative code. You also see that in the dependency handling where you conditionally append and build up a list and then apply that to the target.

Reply:
I'm going to see if the [`target_sources`](https://cmake.org/cmake/help/latest/command/target_sources.html) command will work here

Reply:
I think that if we have a new enough CMake, it should.

Reply:
It appears to work 🎉 The initial version of that command was first introduced in CMake 3.1, and I don't think we're using any of its more advanced functionality added in later versions, so it seems safe for us to rely on here.